src/lib/memory/SimpleMem.cc

```diff
@@ -1,59 +1,60 @@
 #include "simeng/memory/SimpleMem.hh"
 
+#include <algorithm>
+
 namespace simeng {
 namespace memory {
 
 SimpleMem::SimpleMem(size_t size) {
-  memRef = new char[size];
-  std::memset(memRef, 0, size);
-  memory_ = span<char>(memRef, size);
+  memory_ = std::vector<char>(size, '\0');
   memSize_ = size;
 }
 
-SimpleMem::~SimpleMem() { delete[] memRef; };
-
 size_t SimpleMem::getMemorySize() { return memSize_; }
 
-DataPacket* SimpleMem::requestAccess(struct DataPacket* pkt) {
-  if (pkt->type == READ) {
-    struct ReadPacket* rreq = (ReadPacket*)pkt;
-    auto resp = handleReadRequest(rreq);
-    delete rreq;
-    return resp;
-  };
-  struct WritePacket* wreq = (WritePacket*)pkt;
-  auto resp = handleWriteRequest(wreq);
-  delete wreq;
-  return resp;
+DataPacket SimpleMem::requestAccess(struct DataPacket pkt) {
+  if (pkt.type_ == READ_REQUEST) {
+    return handleReadRequest(pkt);
+  } else if (pkt.type_ == WRITE_REQUEST) {
+    return handleWriteRequest(pkt);
+  } else {
+    std::cerr << "[SimEng:SimpleMem] Invalid DataPacket type for requesting "
+                 "access to memory. Requests to memory should either be of "
+                 "type READ_REQUEST or WRITE_REQUEST."
+              << std::endl;
+    return DataPacket(true);
+  }
 };
 
-ReadRespPacket* SimpleMem::handleReadRequest(struct ReadPacket* req) {
-  size_t size = req->size;
-  uint64_t addr = req->address;
-  char* data = new char[size];
-  char* startAddr = memory_.begin() + addr;
-  std::copy(startAddr, startAddr + size, data);
-  return req->makeResponse(size, data);
+DataPacket SimpleMem::handleReadRequest(struct DataPacket req) {
+  size_t size = req.size_;
+  uint64_t addr = req.address_;
+  std::vector<char> data(memory_.begin() + addr, memory_.begin() + addr + size);
+  return req.makeIntoReadResponse(data);
 };
 
-WriteRespPacket* SimpleMem::handleWriteRequest(struct WritePacket* req) {
-  uint64_t address = req->address;
-  size_t size = req->size;
-  const char* data = req->data;
-  std::copy(data, data + size, memory_.begin() + address);
-  return req->makeResponse(size);
+DataPacket SimpleMem::handleWriteRequest(struct DataPacket req) {
+  uint64_t address = req.address_;
+  std::copy(req.data_.begin(), req.data_.end(), memory_.begin() + address);
+  return req.makeIntoWriteResponse();
 };
 
-void SimpleMem::sendUntimedData(char* data, uint64_t addr, size_t size) {
-  std::copy(data, data + size, memory_.begin() + addr);
-  return;
+void SimpleMem::sendUntimedData(std::vector<char> data, uint64_t addr,
+                                size_t size) {
```

Comment:
I have changed the function body to use size, because I want to be safe and avoid any vector memory allocation/iterator based issues we have encountered before. Especially because this is getUntimedData and the vector can be supplied from anywhere